### ISSUE TITLE
Subclasses of os.PathLike should be accepted by joblib.dump

### ISSUE THREAD

**Issue by froody** [joblib/joblib#960]
A subclass of os.PathLike is a valid argument to `open()`, but not to `joblib.dump`, which throws an exception "ValueError: Second argument should be a filename or a file-like object". Any path-like argument accepted by `open()` should also be accepted by `joblib.dump`, including subclasses of `os.PathLike` and instances of `bytes`

**Comment by ogrisel** [comment]
I agree for `os.PathLike` (not so sure for `bytes`). Would you like to contribute a PR.

**Comment by vvegorova** [comment]
`joblib.dump` needs an object passed as `filename` implement the `write` method. `os.PathLike` interface doesn't require `write` to be  implemented.  So, `joblib.dump` can accept only subclasses of  `os.PathLike` that implement `write()`. 

As a solution for this issue, the error message and the doc string (?) can be clarified.

**Comment by froody** [comment]
> `joblib.dump` needs an object passed as `filename` implement the `write` method.

Huh? From the docs it says `joblib.dump` accepts `str` and `pathlib.Path`, both of which are presumably passed to `open` (and neither of which implement `write`). What I'm asking for is for subclasses of `os.PathLike` to be treated like this also.
